### PR TITLE
 wip: split up into core, client and examples

### PR DESCRIPTION
**Pull request by ahjohannessen**
- split project up into three parts:

    - core: contains basic types, protobuf, tcp package code, and otherwise functionality that is not tied to akka that can be reused for alternative implementations.

    - client: contains akka related code and java
      API.

    - examples: contains example code for Scala &
      Java.

- change Java examples, `ReadEventExample` and `WriteEventExample` to use `AbstractActor` instead of deprecated `UntypedActor`.

- add deprecation warning to Java API wrt. usage of publisher methods`.

- bump version to 7.0.0-SNAPSHOT as this builds on joda changes that are breaking.

- add openjdk11 to .travis

- delete `ToCoarsest` as it is available in Scala.

- add `Dependencies` object under project.

### CODE REVIEW

File: client/src/main/scala/eventstore/package.scala [hunk 1]
```diff
@@ -0,0 +1,37 @@
+
+import eventstore.{akka => a}
+import java.{util => ju}
+
+// TODO(AHJ): Remove this package object after 7.1.0
```
Comment (ahjohannessen) [inline client/src/main/scala/eventstore/package.scala]: If it was my call, I would not have this package object at all. I included it in order to make it a bit easier for clients to upgrade. But as this PR builds on a PR that has breaking changes wrt. Joda, we can just as well just break this and put release notes that say specify package changes noted below, as well as the Joda removal.

**Comment by ahjohannessen** [comment]
@t3hnar WDYT?

File: build.sbt [hunk 2]
```diff
@@ -1,108 +1,141 @@
...
-  "-feature",
-  "-unchecked",
-  "-deprecation",
-//  "-Xfatal-warnings",
-  "-Xlint:-missing-interpolator",
-  "-Yno-adapted-args",
-  "-Ywarn-dead-code",
-  "-Ywarn-numeric-widen",
-  "-Xfuture"
+val AkkaVersion            = "2.5.21"
```
[the remark this answers is not in the document]
Comment (ahjohannessen) [reply build.sbt]: Will do. I generally do that in other projects :)

File: build.sbt [hunk 3]
```diff
@@ -1,108 +1,141 @@
...
+    <scm>
+      <url>git@github.com:EventStore/EventStore.JVM.git</url>
+      <connection>scm:git:git@github.com:EventStore/EventStore.JVM.git</connection>
+      <developerConnection>scm:git:git@github.com:EventStore/EventStore.JVM.git</developerConnection>
+    </scm>
+      <developers>
+        <developer>
+          <id>t3hnar</id>
+          <name>Yaroslav Klymko</name>
+          <email>t3hnar@gmail.com</email>
```
[the remark this answers is not in the document]
Comment (ahjohannessen) [reply build.sbt]: Wuhuu :)

File: build.sbt [hunk 4]
```diff
@@ -1,108 +1,141 @@
...
 
-scalacOptions in(Compile, doc) ++= Seq("-groups", "-implicits", "-no-link-warnings")
-
-enablePlugins(ProtobufPlugin)
-
-val AkkaVersion = "2.5.21"
-val AkkaHttpVersion = "10.1.7"
-val ReactiveStreamsVersion = "1.0.2"
-val Specs2Version = "3.8.6" // Because of concurrency issues with specs2 3.8.7+
+lazy val noPublishSettings = Seq(
```
[the remark this answers is not in the document]
Comment (ahjohannessen) [reply build.sbt]: Will change that

File: client/src/main/scala/eventstore/Settings.scala [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+  resolveLinkTos:           Boolean                 = EsDefault.resolveLinkTos,
+  requireMaster:            Boolean                 = EsDefault.requireMaster,
+  readBatchSize:            Int                     = EsDefault.readBatchSize,
+  cluster:                  Option[ClusterSettings] = EsDefault.cluster,
+  http:                     HttpSettings            = HttpSettings.Default,
+  serializationParallelism: Int                     = EsDefault.serializationParallelism,
+  serializationOrdered:     Boolean                 = EsDefault.serializationOrdered,
+  connectionName:           Option[String]          = EsDefault.connectionName,
+  bufferSize:               Int                     = 100000,
+  bufferOverflowStrategy:   a.OverflowStrategy      = a.OverflowStrategy.Fail
```
[the remark this answers is not in the document]
Comment (ahjohannessen) [reply client/src/main/scala/eventstore/Settings.scala]: Not sure I follow. What do you not like?

[the remark this answers is not in the document]
Comment (t3hnar) [reply client/src/main/scala/eventstore/Settings.scala]: a mix of general and akka specific settings

File: client/src/main/scala/eventstore/akka/cluster/cluster.scala [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+import _root_.akka.http.scaladsl.model._
+import _root_.akka.http.scaladsl.unmarshalling.Unmarshal
+import _root_.akka.http.scaladsl.marshallers.sprayjson.SprayJsonSupport
+import eventstore.cluster.ClusterInfo
+
+package object cluster {
+
+  type FutureFunc = InetSocketAddress => Future[ClusterInfo]
+
+  def clusterInfo(implicit system: ActorSystem): FutureFunc = {
```
[the remark this answers is not in the document]
Comment (ahjohannessen) [reply client/src/main/scala/eventstore/akka/cluster/cluster.scala]: There is a `ClusterInfo` case class in the core project. Perhaps `ClusterGossip`?

[the remark this answers is not in the document]
Comment (t3hnar) [reply client/src/main/scala/eventstore/akka/cluster/cluster.scala]: fine with me

[the remark this answers is not in the document]
Comment (t3hnar) [reply client/src/main/scala/eventstore/akka/cluster/cluster.scala]: I often tend to name factor classes as `return type` + `Of` - `ClusterInfoOf`

File: core/src/main/scala/eventstore/util/ToCoarsest.scala [hunk 7]
```diff
@@ -1,11 +1,15 @@
-package eventstore.util
+package eventstore
+package util
 
 import scala.concurrent.duration._
 
 // it is a backport of method introduced in Scala 2.11 - Duration.toCoarsest
-object ToCoarsest {
+private[util] object ToCoarsest {
```
[the remark this answers is not in the document]
Comment (ahjohannessen) [reply core/src/main/scala/eventstore/util/ToCoarsest.scala]: will do

File: client/src/main/scala/eventstore/Settings.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (ahjohannessen) [reply client/src/main/scala/eventstore/Settings.scala]: The real settings are in `eventstore` core project, called `EsSettings` - Those have nothing Akka related.

[the remark this answers is not in the document]
Comment (ahjohannessen) [reply client/src/main/scala/eventstore/Settings.scala]: I think we can migrate off `Settings` later, and make `EsSettings[T](..., extra: T)`, e.g. `EsSettings[Akka](..., extra: Akka)` and have a apply[T](configFn: Config => T) and use a default `apply` that uses `apply[Unit](_ => ())`

File: build.sbt (hunk 2 above)
[the remark this answers is not in the document]
Comment (ahjohannessen) [reply build.sbt]: I do not have so much time now. Could I do this in a separate PR tomorrow?

**Comment by ahjohannessen** [comment]
@t3hnar I have updated according to feedback. I will do a separate PR with dependencies moved into `Dependencies` tomorrow, if that is ok.

File: build.sbt (hunk 2 above)
[the remark this answers is not in the document]
Comment (t3hnar) [reply build.sbt]: Sure

**Comment by ahjohannessen** [comment]
@t3hnar The PR is updated to include `Dependencies` and `skip in publish := true`

**Pull request merged**